miner/currency_blocklist.go

```diff
@@ -18,24 +18,89 @@ type AddressBlocklist struct {
 }
 
 func NewAddressBlocklist() *AddressBlocklist {
-	return &AddressBlocklist{
+	bl := &AddressBlocklist{
 		mux:                          &sync.RWMutex{},
 		currencies:                   map[common.Address]*types.Header{},
 		headerEvictionTimeoutSeconds: EvictionTimeoutSeconds,
 		oldestHeader:                 nil,
+		disabledCurrencies:           make(map[common.Address]struct{}),
 	}
+	return bl
+}
+
+// BlocklistFeeCurrencies returns the current blocklist, a set of fee currency
+// addresses mapped to their expiry Unix timestamp. Note that the parameter
+// 'includeDisabled' determines if the map should contain fee currencies for
+// which blocking has been manually disabled.
+func (b *AddressBlocklist) Blocklist(includeDisabled bool) map[common.Address]uint64 {
+	b.mux.RLock()
+	defer b.mux.RUnlock()
+	result := make(map[common.Address]uint64, len(b.currencies))
+	for currency, addedHeader := range b.currencies {
+		result[currency] = addedHeader.Time + b.headerEvictionTimeoutSeconds
+	}
+	return result
+}
+
+// EnableBlocking enables blocking on the given currencies, for currencies that
+// are already enabled this is a no-op.
+func (b *AddressBlocklist) EnableBlocking(currencies []common.Address) {
+	b.mux.Lock()
+	defer b.mux.Unlock()
+	for _, currency := range currencies {
+		b.disabledCurrencies[currency] = struct{}{}
+	}
+}
+
+// DisableBlocking disables blocking on the given currencies, for currencies that
+// are already disabled this is a no-op.
+func (b *AddressBlocklist) DisableBlocking(currencies []common.Address) {
+	b.mux.Lock()
+	defer b.mux.Unlock()
+	for _, currency := range currencies {
+		delete(b.disabledCurrencies, currency)
+	}
+}
+
+// DisabledCurrencies returns the currencies for which blocking is currently
+// manually disabled.
+func (b *AddressBlocklist) DisabledCurencies() []common.Address {
+	b.mux.RLock()
+	defer b.mux.RUnlock()
+	result := make([]common.Address, 0, len(b.disabledCurrencies))
+	for currency := range b.disabledCurrencies {
+		result = append(result, currency)
+	}
+	return result
+}
+
+// BlockingEnabled returns whether blocking is enabled for the given currency.
+func (b *AddressBlocklist) BlockingEnabled(currency common.Address) bool {
+	b.mux.RLock()
+	defer b.mux.RUnlock()
+	_, disabled := b.disabledCurrencies[currency]
+	return !disabled
 }
 
+// FilterAllowlist returns allowlist with any blocked and not disabled
+// currencies removed. It accepts the latest header so that it may update the
+// blocklist by evicting any blocked fee currencies that have exceeded their
+// timeout before it proceeds to filter the allowlist.
 func (b *AddressBlocklist) FilterAllowlist(allowlist common.AddressSet, latest *types.Header) common.AddressSet {
 	b.mux.RLock()
 	defer b.mux.RUnlock()
 
 	filtered := common.AddressSet{}
 	for a := range allowlist {
-		if !b.isBlocked(a, latest) {
+		_, disabled := b.disabledCurrencies[a]
+		if !b.isBlocked(a, latest) && !disabled {
```

Comment:
We want to return not blocked or blocking disabled currencies, right?
```suggestion
		if !b.isBlocked(a, latest) || disabled {
```

Reply:
Yes that's right